File: tests/liregress_file.cpp
#include "minerva.h"

#include <cstring>
#include <fstream>

using namespace std;
using namespace minerva;

void Print(float* ptr) {
  for(int i = 0; i < 10; ++i)
    cout << ptr[i] << " ";
  cout << endl;
}

int main(int argc, char** argv) {
  MinervaSystem::Instance().Initialize(argc, argv);
  int n = 10; // num samples
  int k = 8; // num features
  //NArray x = NArray::Randn({n, k}, 0.0, 1.0, {1, 2});
  //NArray y = NArray::Randn({n, k}, 0.0, 1.0, {1, 2});
  //NArray theta = NArray::Randn({k, k}, 0.0, 1.0, {2, 2});
  SimpleFileLoader loader;
  NArray x = NArray::LoadFromFile({n, k}, "x.dat", &loader, {1, 1});
  NArray y = NArray::LoadFromFile({n, k}, "y.dat", &loader, {1, 1});
  NArray theta = NArray::LoadFromFile({k, k}, "theta.dat", &loader, {1, 1});
  float alpha = 0.5; // lr
  int epoch = 2;
  for(int i = 0; i < epoch; ++i) {
    NArray error = x * theta - y;
    theta = theta - alpha * x.Trans() * error;
  }
  FileFormat format; format.binary = false;
  theta.ToFile("theta_trained.txt", format);
  format.binary = true;
  theta.ToFile("theta_trained.dat", format);
  return 0;
}

File: owl/apps/liregress_file.py
#!/usr/bin/env python

import sys

import owl
from owl import SimpleFileLoader
from owl import FileFormat


def main(args=sys.argv[1:]):
    owl.initialize(sys.argv)
    n = 10
    k = 8
    loader = SimpleFileLoader()
    x = owl.load_from_file((n, k), 'x.dat', loader, (1, 1))
    y = owl.load_from_file((n, k), 'y.dat', loader, (1, 1))
    theta = owl.load_from_file((k, k), 'theta.dat', loader, (1, 1))
    alpha = 0.5
    epoch = 2
    for i in xrange(epoch):
        error = x * theta - y
        theta = theta - alpha * x.trans() * error  # TODO: x.T
    format = FileFormat()
    format.binary = False
    theta.to_file('theta_trained.txt', format)
    format.binary = True
    theta.to_file('theta_trained.dat', format)
    print 'linear regression finished.'


if __name__ == '__main__':
    sys.exit(main())
